ext/standard/math.c

```diff
@@ -249,6 +249,108 @@ PHPAPI double _php_math_round(double value, int places, int mode) {
 }
 /* }}} */
 
+/* {{{ _php_math_round_long */
+/*
+ * Rounds a zend_long to a certain number of decimal places in a certain rounding
+ * mode. For the specifics of the algorithm, see https://wiki.php.net/rfc/integer-rounding
+ */
+PHPAPI zend_result _php_math_round_long(zend_long value, int places, int mode, zend_long *result) {
```

Comment:
Frankly I find this implementation incomprehensible. I've specifically refactored the floating point implementation to leverage a switch statement with *very* simple logic to make the code easier to verify for correctness in #12220.

This pattern should likely also be applied here. It also doesn't help that the logic is effectively duplicated for the “above zero” and “below zero” cases.

Reply:
Hi Tim,
This is a first working version actually started before #12220 and as such I was following the previous style.
But you are right this can and should be improved. If the RFC gets accepted I'll update the PR according.